### ISSUE TITLE
How to handle borrow checking in data races caused by callbacks?

### ISSUE THREAD

**Issue by alshdavid** [alshdavid/BorrowScript#3]
In the example below we have an http server that runs a callback function when any request is received. The callback function will increment a counter. The callback function can be scheduled on multiple threads and will therefore put the counter at a risk of data races. 

It's important to prevent compilation in this scenario - however this scenario does not violate the borrow checker's rules.

Assuming [the syntax described in this issue to pass variables from an outer scope in](https://github.com/alshdavid/TypeScriptBC/issues/2) then we would pass write access into the handler like:

```typescript
let counter: number = 0

server.handle(function(move c: http.Context)[write counter]: void {
  counter.increment()

  c.response.setStatus(200)
  c.response.send()
})
```
Obviously we would want to use a `Mutex` to wrap `counter`, but the compiler needs to catch this and prevent compilation. What should happen here?

**Comment by SuperSonicHub1** [comment]
I assume the borrow checker allows us at least some static analysis, so assuming that:

- Find all instances of callbacks through checking function signatures
- Iterate through all of the bodies of the callbacks
- Find any instances of writing
- If those instances of writing aren't using a lock, raise a compiler error

**Comment by alexvargasbenamburg** [comment]
I second @SuperSonicHub1 that is a good idea in how to implement it.

**Comment by alshdavid** [comment]
That makes sense. This would also need to account for `read` and `move` operations though. Move doesn't make sense in the context of a callback because there is an indeterminate number of callback instances spawned - potentially at the same time (race condition).

I'm digging into how Rust deals with this for inspiration (seeing as TSBC is basically simplified rust, hey)

**Comment by alshdavid** [comment]
Rust uses the `Send` and `Sync` traits to mark a value as "safe to transfer over threads". The compiler will automatically determine if the `Sync` trait is required by following calls down to `thread::spawn`.

To give us read and write access to a value across threads, we use an `Arc` and `Mutex`. `Arc` will allow us to make multiple "smart pointers" to a source, counting how many pointers are lent out and will drop when the pointers reach 0 where the `Mutex` will allow data to be mutated from multiple threads.

These are provided by the standard library but they are well tested wrappers around "unsafe" code. e.g. we _know_ that a Mutex protects against concurrent writes so it's okay to lend it out to do so.

So something like this in Rust:
```rust
fn main() {
  // Arc implements Send, Mutex impairments Send and Sync
  let counter = Arc::new(Mutex::new(5));

  // We create a new smart pointer reference to our mutex
  let counterRef = Arc::clone(&counter);

  // move our counterRef into the callback
  http_get("/foo",  &(move || {
    // Lock the mutex
    let mut counter = counterRef.lock().unwrap();
    *counter = *counter + 1;
    println!("/foo: {}", c);
  }));
}
```

So seeing as we are simplifying things here, let's drop `Arc` and ensure the internal implementation of `Mutex` handles that.

```typescript
function main() {
   const counter = new Mutex(0)

   const counterRef = counter.clone() 
   httpGet('/foo', function()[move counterRef] {
      let counter = counterRef.lock() // mutable reference to the counter value
      counter.increment() // we must mutate it, not reassign it
      console.log(counter)
      counterRef.unlock()
   })
}
```

Perhaps this is where we could see the `copy` operator come in handy where it handles cloning the Mutex reference.
```typescript
function main() {
   const counterMutex = new Mutex(0)

   httpGet('/foo', function()[copy counterMutex] {
      let counter = counterMutex.lock()
      counter.increment()
      console.log(counter)
      counterRef.unlock()
   })
}
```

**Comment by alshdavid** [comment]
I am calling this expansion to function declaration "ownership gates" as they act as a gate to allow parameters in an outer scope to be moved into a target function.

I think we can only support `move` and maybe `copy` (to shorthand cloning a `Mutex`)

Here is a TypeScript concept

```typescript
let value = 0

setTimeout(() => {
  console.log(value)
})
```

Let's assume that `setTimeout` executes on another thread.  Such access is actually not a thing in Rust (threaded or otherwise). The `value` variable must be explicitly added to the nested scope.

Here is a Rust example
```rust
fn main() {
  // Arc + Mutex allows us to send something into another thread, getting around the "1 writer" rule
  let counter = Arc::new(Mutex::new(5));

  // When you clone an arc reference, you get a copy of the pointer, this is known as a smart pointer
  let counterRef = Arc::clone(&counter);

  // This is a Rust lambda fn, the move keyword here just says "Move ownership of anything I use from my outer scope"
  set_timeout(&(move || {
    // Get a mutable reference to the value within the mutex
    let mut counter = counterRef.lock().unwrap();
    *counter = *counter + 1;
    println!("/foo: {}", c);
    // Mutex is locked automatically on drop
  }));
 
  // Because counterRef  was moved into the callback, we can actually redeclare it in 
  // the main code block as it's actually really been moved out of this block entirely
  let counterRef = Arc::clone(&counter);
}
```

To achieve a simplified form of this syntax in BorrowScript we have this:

```typescript
import { Mutex, setTimeout } from '@std/sync'

async function main() {
  const counterRef = new Mutex(0)
  const counterRef1 = copy counterRef

  setTimeout (()[ move counterRef1 ] => {
    let value = counterRef1.lock()
    value.increment()
  })
}
```

Which I am experimenting with further simplifying by
```typescript
import { Mutex, setTimeout } from '@std/sync'

async function main() {
  const counterRef = new Mutex(0)

  setTimeout(()[ copy counterRef ] => { // create a new variable in this scope called "counterRef" from its copy
    let value = counterRef.lock()
    value.increment()
  })
}
```

So basically for functions that want to pass variables from their outer scope in, we have "ownership gates" in the function declaration syntax

```typescript
function<T>(params)[copy|move variables] { code }
```

**Comment by alshdavid** [comment]
`server.handle` would internally distribute its requests across a queue, this means it could be multithreaded if it's queue is allocated multiple threads.

The issue with callbacks is about ownership. A callback cannot exist without some sort of syntax indicating ownership transferal. So in your example

```typescript
// ...
  innewThread(()=>{
    counter.increment() // <-- "counter" is not owned by the current scope
// ...
```
For this to work you need to say "from the outer scope, move the ownership of following into this scope"
```
innewThread(()[move counter]=>{
    counter.increment()
```
But at that point, `server.handle` is already multi-threaded so there is no point in `innerThread` existing - unless you want a thread inside a thread.
